### PR TITLE
Clean orphaned (missing) nodes

### LINKED ISSUE #222: Possibility to cleanup orphaned nodes
## Expected Behavior
Having a button to delete all deleted services and hosts in one step.

## Current Behavior
You have to delete all deleted services and hosts one by one.

## Context
Hosts and services get deleted, but they are still part of the businessprocesses. While you can easily recognize by the color which ones are to be deleted, you have to delete them one by one which can be at least time-consuming. Making this a one-click action, makes maintenance of businessprocesses more easy.

ref/NC/590017

### PR DESCRIPTION
**Pull request by sukhwinder33445**
resolves #222 

## Blocked by:
- https://github.com/Icinga/icingaweb2-module-businessprocess/pull/358

### CODE REVIEW

**Comment by sukhwinder33445** [comment]
Changes note:
When the process contains linked nodes (in which nodes are missing), 
 only the `Show` link is displayed. If there are no linked missing nodes, but the process contains direct missing nodes, the `Cleanup` link is displayed.

This is necessary because the process cannot differentiate between direct missing nodes and linked missing nodes.

**Review by raviks789 — DISMISSED** [review]

File: application/forms/CleanupNodeForm.php [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            'label'         => $this->translate('Cleanup all missing nodes'),
+            'description'   => $this->translate('Remove all missing nodes from config')
+        ]);
+
+        if ($this->getSentValue('cleanup_all') !== '1') {
+            $this->addElement('multiselect', 'nodes', [
+                    'label'        => $this->translate('Select nodes to cleanup'),
+                    'required'     => true,
+                    'size'         => 8,
+                    'style'        => 'width: 25em',
```
Comment (nilmerg) [inline application/forms/CleanupNodeForm.php:47]: nah, csp and so on, you know ;)

File: application/forms/CleanupNodeForm.php [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+class CleanupNodeForm extends QuickForm
+{
+    /** @var MonitoringBackend|IcingaDbConnection */
+    protected $backend;
+
+    /** @var BpConfig */
+    protected $bp;
+
+    /** @var BpNode */
+    protected $parentNode;
```
Comment (nilmerg) [inline application/forms/CleanupNodeForm.php:22]: seems to be unused

File: application/clicommands/CleanupCommand.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    public function cleanupAction()
+    {
+        $configNames = (array) $this->params->shift() ?: $this->storage->listAllProcessNames();
+        $missingChildren = $this->listAllMissingChildren($configNames);
+        $removedNodes = [];
+
+        foreach ($missingChildren as $nodeName) {
+            foreach ($configNames as $configName) {
+                if ($this->storage->hasProcess($configName)) {
+                    $bp = $this->storage->loadProcess($configName);
```
Comment (nilmerg) [inline application/clicommands/CleanupCommand.php:50]: `listAllMissingChildren` did all of this already. Cache it. Use it. Or just merge the two methods, or use a generator. Just be more efficient please.

File: application/clicommands/CleanupCommand.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                    }
+                }
+            }
+        }
+
+        if (count($removedNodes)) {
+            echo sprintf('Removed following %d missing nodes successfully:', count($removedNodes)) . "\n";
+            foreach ($removedNodes as $node) {
+                echo $node . "\n";
+            }
```
Comment (nilmerg) [inline application/clicommands/CleanupCommand.php:81]: include the config name. If none is passed, there's no way of knowing which config was involved later on.

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: application/clicommands/CleanupCommand.php [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            foreach (array_keys($bp->getMissingChildren()) as $missingNode) {
+                    $node = $bp->getNode($missingNode);
+                    $remove = new NodeRemoveAction($node);
+                    if ($remove->appliesTo($bp)) {
+                       $remove->applyTo($bp);
+                        $removedNodes[] = $node->getName();
+                    }
+
+                    $this->storage->storeProcess($bp);
+                    $bp->clearAppliedChanges();
```
Comment (nilmerg) [inline application/clicommands/CleanupCommand.php:84]: Please check the indentation.

File: application/clicommands/CleanupCommand.php [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+            }
+
+            if (! empty($removedNodes)) {
+                echo sprintf('Removed following %d missing node(s) from %s.conf successfully:',
+                    count($removedNodes),
+                    $configName
+                    );
```
Comment (nilmerg) [inline application/clicommands/CleanupCommand.php:91]: Please check the indentation.

File: application/clicommands/CleanupCommand.php [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+     */
+    public function cleanupAction(): void
+    {
+        $configNames = (array) $this->params->shift() ?: $this->storage->listAllProcessNames();
+
+        $removedNodes = null;
+        foreach ($configNames as $configName) {
+            if (! $this->storage->hasProcess($configName)) {
+                Logger::error('Failed to scan the %s.conf file for missing nodes. File not found', $configName);
+                $removedNodes = false;
```
Comment (nilmerg) [inline application/clicommands/CleanupCommand.php:50]: Don't do this. There is no reason not to log that no missing nodes are found. If no config is readable, of course none are found.

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: application/controllers/ProcessController.php [hunk 8]
```diff
@@ -366,6 +368,64 @@ protected function showHints(BpConfig $bp)
...
+                    $missingLinkedNodes = array_keys($process->getMissingChildren());
+                    $link = Url::fromPath('businessprocess/process/show')
+                        ->addParams(['config' => $process->getConfigName()]);
+
+                    $ul->add(Html::tag(
+                        'li',
+                        [
+                            TemplateString::create(
+                                tp(
+                                    'Linked node %s has %d missing child node: {{#link}}Show{{/link}}',
```
Comment (nilmerg) [inline application/controllers/ProcessController.php:387]: ~~%d~~one

File: application/controllers/ProcessController.php [hunk 9]
```diff
@@ -366,6 +368,64 @@ protected function showHints(BpConfig $bp)
...
+                                    'Linked node %s has %d missing child nodes: {{#link}}Show{{/link}}',
+                                    count($missingLinkedNodes)
+                                ),
+                                $process->getAlias(),
+                                count($missingLinkedNodes),
+                                ['link' => new Link(null, (string)$link)]
```
Comment (nilmerg) [inline application/controllers/ProcessController.php:393]: Space after casts

File: application/controllers/ProcessController.php [hunk 10]
```diff
@@ -366,6 +368,64 @@ protected function showHints(BpConfig $bp)
...
+
+            $link = Url::fromPath('businessprocess/process/show')
+                ->addParams(['config' => $this->bp->getName(), 'action' => 'cleanup']);
+
+            $ul->add(Html::tag(
+                'li',
+                [
+                    TemplateString::create(
+                        t('{{#link}}Cleanup{{/link}} %d missing nodes: %s'),
+                        ['link' => new Link(null, (string)$link)],
```
Comment (nilmerg) [inline application/controllers/ProcessController.php:418]: space after casts

File: application/forms/CleanupNodeForm.php [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+    protected $bp;
+
+    /** @var SessionNamespace */
+    protected $session;
+
+    public function setup()
+    {
+        $this->addHtml(Html::tag(
+            'h2',
+            FormattedString::create($this->translate('Cleanup missing nodes'))
```
Comment (nilmerg) [inline application/forms/CleanupNodeForm.php:29]: That's not a formatted string, but just a string.

Comment (nilmerg) [reply application/forms/CleanupNodeForm.php:29]: My previous comment was connected to this being a string, not `ValidHtml`.

**Review by nilmerg — CHANGES_REQUESTED** [review]
Also, please rebase

**Review by nilmerg — APPROVED** [review]

**Pull request merged**